ttnn/cpp/ttnn/operations/moreh/moreh_sgd/device/moreh_sgd_device_operation.cpp

```diff
@@ -0,0 +1,114 @@
+
+namespace ttnn::operations::moreh::moreh_sgd {
+void MorehSgdOperation::validate_inputs(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    auto& params_in = tensor_args.param_in;
+    auto& grad = tensor_args.grad;
+
+    tt::operations::primary::check_tensor(params_in, "moreh_sgd", "params_in");
+    tt::operations::primary::check_tensor(grad, "moreh_sgd", "grad");
+
+    if (tensor_args.momentum_buffer_in) {
+        tt::operations::primary::check_tensor(*tensor_args.momentum_buffer_in, "moreh_sgd", "momentum_buffer_in");
+    }
+
+    if (tensor_args.param_out.has_value()) {
+        tt::operations::primary::check_tensor(tensor_args.param_out.value(), "moreh_sgd", "param_out");
+    }
+
+    if (tensor_args.momentum_buffer_out.has_value()) {
+        tt::operations::primary::check_tensor(
+            tensor_args.momentum_buffer_out.value(), "moreh_sgd", "momentum_buffer_out");
+    }
+}
+
+MorehSgdOperation::program_factory_t MorehSgdOperation::select_program_factory(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    return ProgramFactory{};
+};
+
+void MorehSgdOperation::validate_on_program_cache_miss(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    validate_inputs(operation_attributes, tensor_args);
+};
+
+void MorehSgdOperation::validate_on_program_cache_hit(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    validate_inputs(operation_attributes, tensor_args);
+};
+
+MorehSgdOperation::shape_return_value_t MorehSgdOperation::compute_output_shapes(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    auto input_tensor_shape = tensor_args.param_in.get_shape();
+
+    return {input_tensor_shape, input_tensor_shape};
+};
+
+MorehSgdOperation::tensor_return_value_t MorehSgdOperation::create_output_tensors(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    const auto& output_shapes = compute_output_shapes(operation_attributes, tensor_args);
+    auto dtype = tensor_args.param_in.get_dtype();
+    Layout layout{Layout::TILE};
+    auto device = tensor_args.param_in.device();
+
+    std::vector<std::optional<Tensor>> ret;
+
+    if (tensor_args.param_out.has_value()) {
+        ret.push_back(tensor_args.param_out.value());
+    } else {
+        ret.push_back(create_device_tensor(
+            output_shapes.at(0).value(), dtype, layout, device, operation_attributes.param_out_memory_config));
+    }
+
+    if (tensor_args.momentum_buffer_out.has_value()) {
+        ret.push_back(tensor_args.momentum_buffer_out.value());
+    } else {
+        ret.push_back(create_device_tensor(
+            output_shapes.at(1).value(),
+            dtype,
+            layout,
+            device,
+            operation_attributes.momentum_buffer_out_memory_config));
+    }
```

Comment:
`momentum_buffer_out` only produces output when the momentum is not zero.

Reply:
I resolved this once, but Duong mistakenly force-pushed and removed the changes. When you make changes in `device_operation.cpp`, you also need to update `program_factory.cpp`
@DuongQLee 

Reference: https://github.com/tenstorrent/tt-metal/blob/85b6ec1c842f1ce0cfa8f9e9d7ea29f994a16920/ttnn/cpp/ttnn/operations/moreh/moreh_sgd/device/moreh_sgd_program_factory.cpp

Reply:
I have resolved this.